src/validation/Validator.ts

```diff
@@ -10,65 +10,41 @@ import * as validator from "validator";
  * Validator performs validation of the given object based on its metadata.
  */
 export class Validator {
-
-    // -------------------------------------------------------------------------
-    // Private Properties
-    // -------------------------------------------------------------------------
-
-
-    /**
-     * Performs validation of the given object based on decorators or validation schema.
-     * Common method for `validateOrReject` and `validate` methods.
-     */
-    private coreValidate(objectOrSchemaName: Object|string, objectOrValidationOptions: Object|ValidationOptions, maybeValidatorOptions?: ValidatorOptions): Promise<ValidationError[]> {
-        const object = typeof objectOrSchemaName === "string" ? objectOrValidationOptions as Object : objectOrSchemaName as Object;
-        const options = typeof objectOrSchemaName === "string" ? maybeValidatorOptions : objectOrValidationOptions as ValidationOptions;
-        const schema = typeof objectOrSchemaName === "string" ? objectOrSchemaName as string : undefined;
-
-        const executor = new ValidationExecutor(this, options);
-        const validationErrors: ValidationError[] = [];
-        executor.execute(object, schema, validationErrors);
-
-        return Promise.all(executor.awaitingPromises).then(() => {
-            return executor.stripEmptyErrors(validationErrors);
-        });
-    }
-
     // -------------------------------------------------------------------------
     // Public Methods
     // -------------------------------------------------------------------------
 
     /**
      * Performs validation of the given object based on decorators used in given object class.
      */
-    validate(object: Object, options?: ValidatorOptions): Promise<ValidationError[]>;
+    validate(object: Record<string, any>, options?: ValidatorOptions): Promise<ValidationError[]>;
```

Comment:
Are we sure `Record<string, any>` is same as `Object`? This is change in our API so I want to be sure.
I think it is ok but I want extra check. @satanTime could you please check this?

Reply:
almost, but key can also be a number and a symbol, the best way would be `keyof any` instead of `stream`. I'm checking the changes and prepare a feedback soon.

Reply:
The error message is:
```
class-validator/src/validation/Validator.ts
  37:30  error  Don't use 'Object' as a type. Use Record<string, any> instead  @typescript-eslint/ban-type
```

but the definition of `Record` is
```typescript
/**
 * Construct a type with a set of properties K of type T
 */
type Record<K extends keyof any, T> = {
    [P in K]: T;
};
```

To be honest I don't know people who are using objects with number / symbol keys, but if it's supported in an object would be good to see it in the lib too.
```typescript
Record<keyof any, any>
```

Reply:
Maybe use `object` (lowercase first "o") - https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#object-type

Reply:
IMO works too, a replacement of `Object` with `{[key: keyof any]: any}` or `Record` etc. for me is just a hotfix of linting and in case of general libraries such as class-validator to be precise isn't so easy.
As I understand from documentation `Object` is kind of `any` because it matches primitives when `object` is an object as we understand it https://take.ms/RYUk6
if we expect a normal object - then `object` is the perfect match considering the `null` case.

Reply:
```ts
declare function original(o: Object): void;
declare function objectType(o: object): void;
declare function record(o: Record<any, any>): void;

class TestClass { }

// ok case
original({ a: 1, 1: 2 }); // OK
objectType({ a: 1, 1: 2 }); // OK
record({ a: 1, 1: 2 }); // OK

original(new TestClass()); // OK
objectType(new TestClass()); // OK
record(new TestClass()); // OK

original(Object.create({})); // OK
objectType(Object.create({})); // OK
record(Object.create({})); // OK

original(Object.create(null)); // OK
objectType(Object.create(null)); // OK
record(Object.create(null)); // OK

original(null); // Error
objectType(null); // Error
record(null); // Error

original(undefined); // Error
objectType(undefined); // Error
record(undefined); // Error

original(Symbol("key")); // OK
objectType(Symbol("key")); // Error
record(Symbol("key")); // Error

original(true); // OK
objectType(true); // Error
record(true); // Error

original('asd'); // OK
objectType('asd'); // Error
record('asd'); // Error

original(1); // OK
objectType(1); // Error
record(1); // Error

original(() => false); // OK
objectType(() => false); // OK
record(() => false); // OK
```
Looks like `Record<any, any>` is equivalent for `object`. However this looks like public API change (we should move it into 0.13.0 release) **from my point of view is this fix because we want only objects as input for validation 😄**

Reply:
agree, make sense. For me Record is too long to say any object. Anyway `Object` doesn't support `null`, so `object` looks like the best match.

Reply:
Changed Record<string, any> to object.

Reply:
I'd like to reopen the discussion on Record<string, any> vs object :)

There are some subtle differences that may impact class-validator.
object is a primitive type, which means we cannot use it with generics

Perhaps Record<string, any> is the correct usage here?

```
declare function genericRecord<T extends Record<string, any>>(o: T): void;
declare function genericObject<T extends object>(o: T): void;

function genericRecordWrapper<T>(o: T): void {
    genericRecord(o); // OK
}

function genericObjectWrapper<T>(o: T): void {
    genericObject(o); // ERROR
   // Argument of type 'T' is not assignable to parameter of type 'object' (2345)
}
```

Reply:
perhaps Record<string | Symbol, any> is better?

Reply:
the right type for keys is `keyof any`, but your point is interesting, I'll check it deeper tonight too.

Reply:
Are you sure?
From the Moz link above:
"Property names are string or Symbol. Any other value, including a number, is coerced to a string."

Reply:
Moz is about javascript when we discuss typescript here and in its case number and string are different types. Check source code of Record and you can see `keyof any` there. I would rely on knowledge of typescript's team.

Reply:
Ah, good point. I understand. So perhaps Record<any, any> is the best solution?

Reply:
I've been thinking about this some more and I've realized that I was wrong.
object = any non primitive type
So my earlier code was wrong/badly typed.
```
declare function genericRecord<T extends Record<string, any>>(o: T): void;
declare function genericObject<T extends object>(o: T): void;

function genericRecordWrapper<T>(o: T): void {
    genericRecord(o); // OK
}

// This is now a correct definition. It does not allow primitive types
function genericObjectWrapper<T extends object>(o: T): void {
// function genericObjectWrapper<T>(o: T): void {
    genericObject(o); // OK
}
```
So object *is* correct usage for class-validator :)

Reply:
Ah, call, I almost started the investigation. Then glad to hear and looking forward to see your changes in merged :)